tools/autograd/gen_variable_type.py

```diff
@@ -646,20 +651,36 @@ def save_variables(
                 stmts.append('}')
         return stmts
 
-    def emit_dispatch_call(f: NativeFunction, input_base: str, unpacked_args: Sequence[str]) -> str:
+    def emit_dispatch_call(f: NativeFunction, input_base: str, unpacked_args: Sequence[str], *, is_view_call: bool) -> str:
         """ Dispatch call via function in a namespace or method on Tensor."""
-        if Variant.function in f.variants:
-            call = CALL_DISPATCH_VIA_NAMESPACE.substitute(
+        dispatcher_sig = DispatcherSignature.from_schema(f.func)
+        dispatcher_exprs = dispatcher_sig.exprs()
+
+        if is_view_call:
+            # View replay functions use the standard Dispatcher::call API, since they are not hot-path.
+            if Variant.function in f.variants:
+                call = CALL_DISPATCH_VIA_NAMESPACE.substitute(
+                    api_name=cpp.name(
+                        f.func,
+                        faithful_name_for_out_overloads=True,
+                    ),
+                    unpacked_args=unpacked_args)
+            else:
+                call = CALL_DISPATCH_VIA_METHOD.substitute(
+                    api_name=cpp.name(f.func),
+                    var=input_base,
+                    unpacked_method_args=unpacked_args[1:])
+        else:
+            # code-generated autograd kernels plumb and recompute dispatch keys directly through the kernel for performance.
+            # Ops also always have a function variant of the redispatch API.
+            # See Note [Plumbing Keys Through The Dispatcher] for details.
+            dispatch_key_set = 'ks & c10::after_autograd_keyset'
```

Comment:
Is the `c10::after_autograd_keyset` actually needed here since we set the TLS guard? 
I guess the answer is yes, but then it feels dangerous that we have to set both this and the TLS guard (one of the two can easily be forgotten).

Reply:
yeah that's right. I agree that needing to set it in two places is dangerous. Just to lay out why:
* The TLS guard is the important one for correctness, ensuring that we never hit another autograd kernel from this re-dispatch or from any base op calls
* The use of `c10::after_autograd_keyset` is more for performance than correctness, since pre-computing the keyset means the dispatcher doesn't need to compute it itself and read from TLS. (Although it also ensures that `RecordFunction` isn't called a second time in the dispatch chain)

Reply:
I am curious, what would happen if I do the following:
- Set the TLS properly
- Call into the redispatch API
- pass only `ks` to it (forget to do the ` & c10::after_autograd_keyset`)

Will that infinite recurse into the autograd kernel?

Reply:
Yep, that's exactly right. Since calling `Dispatcher::redispatch()` basically says to the dispatcher: "don't waste your time computing the full DispatchKeySet from arguments, TLS, etc- just use exactly the DispatchKeySet I gave you". TLS keys then only matter when you have an op who's backend kernel calls out to other base ops.

So this API improves dispatcher perf, at the cost of making it a little easier to shoot yourself in the foot. I think that's okay here, since (a) we're only using it in internal redispatching kernels where we have full control and we care about core perf, e.g. it's not something that the average user will be exposed to, and (b) an infinite recursion like that will fail fast from CI, etc.

Reply:
That sounds good.
Yes, if the only way this goes wrong is by reliably blowing up in a big boom, I'm fine with it!

Also we might want to capture this discussion in a comment somewhere. Not sure where is best though.

Reply:
I don't explicitly mention the way you can shoot yourself in the foot, but the API is documented here: https://github.com/pytorch/pytorch/blob/7363da7c57729d70532b647935dd6a9e68be612a/aten/src/ATen/core/dispatch/Dispatcher.h#L148